File: packages/google_mlkit_commons/android/src/main/java/com/google_mlkit_commons/GenericModelManager.java
package com.google_mlkit_commons;

import com.google.mlkit.common.model.DownloadConditions;
import com.google.mlkit.common.model.RemoteModel;
import com.google.mlkit.common.model.RemoteModelManager;

import io.flutter.plugin.common.MethodCall;
import io.flutter.plugin.common.MethodChannel;

public class GenericModelManager {
    private static final String DOWNLOAD = "download";
    private static final String DELETE = "delete";
    private static final String CHECK = "check";

    public interface CheckModelIsDownloadedCallback {
        void onCheckResult(Boolean isDownloaded);

        void onError(Exception e);
    }

    public RemoteModelManager remoteModelManager = RemoteModelManager.getInstance();

    public void manageModel(final RemoteModel model, final MethodCall call, final MethodChannel.Result result) {
        String task = call.argument("task");

        if (task == null) {
            result.notImplemented();
            return;
        }

        switch (task) {
            case DOWNLOAD:
                boolean isWifiReqRequired = call.argument("wifi");
                DownloadConditions downloadConditions;
                if (isWifiReqRequired)
                    downloadConditions = new DownloadConditions.Builder().requireWifi().build();
                else
                    downloadConditions = new DownloadConditions.Builder().build();
                downloadModel(model, downloadConditions, result);
                break;
            case DELETE:
                deleteModel(model, result);
                break;
            case CHECK:
                isModelDownloaded(
                        model,
                        new CheckModelIsDownloadedCallback() {
                            @Override
                            public void onCheckResult(Boolean isDownloaded) {
                                result.success(isDownloaded);
                            }

                            @Override
                            public void onError(Exception e) {
                                result.error("error", e.toString(), null);
                            }
                        }
                );
                break;
            default:
                result.notImplemented();
        }
    }

    public void downloadModel(RemoteModel remoteModel, DownloadConditions downloadConditions, final MethodChannel.Result result) {
        isModelDownloaded(
                remoteModel,
                new CheckModelIsDownloadedCallback() {
                    @Override
                    public void onCheckResult(Boolean isDownloaded) {
                        if (isDownloaded) {
                            result.success("success");
                            return;
                        }

                        remoteModelManager.download(remoteModel, downloadConditions)
                                .addOnSuccessListener(aVoid -> result.success("success"))
                                .addOnFailureListener(e -> result.error("error", e.toString(), null));
                    }

                    @Override
                    public void onError(Exception e) {
                        result.error("error", e.toString(), null);
                    }
                }
        );
    }

    public void deleteModel(RemoteModel remoteModel, final MethodChannel.Result result) {
        isModelDownloaded(remoteModel, new CheckModelIsDownloadedCallback() {
            @Override
            public void onCheckResult(Boolean isDownloaded) {
                if (!isDownloaded) {
                    result.success("success");
                    return;
                }
                remoteModelManager.deleteDownloadedModel(remoteModel)
                        .addOnSuccessListener(aVoid -> result.success("success"))
                        .addOnFailureListener(e -> result.error("error", e.toString(), null));
            }

            @Override
            public void onError(Exception e) {
                result.error("error", e.toString(), null);
            }
        });
    }

    public void isModelDownloaded(RemoteModel model, CheckModelIsDownloadedCallback callback) {
        try {
            remoteModelManager.isModelDownloaded(model)
                    .addOnFailureListener(callback::onError)
                    .addOnSuccessListener(callback::onCheckResult);
        } catch (Exception e) {
            callback.onError(e);
        }
    }
}

File: packages/google_mlkit_commons/android/src/main/kotlin/com/google_mlkit_commons/GenericModelManager.kt
package com.google_mlkit_commons

import com.google.mlkit.common.model.DownloadConditions
import com.google.mlkit.common.model.RemoteModel
import com.google.mlkit.common.model.RemoteModelManager
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import java.lang.reflect.Method

class GenericModelManager {
    interface CheckModelIsDownloadedCallback {
        fun onCheckResult(isDownloaded: Boolean?)

        fun onError(e: Exception)
    }

    val remoteModelManager: RemoteModelManager = RemoteModelManager.getInstance()

    companion object {
        private const val DOWNLOAD = "download"
        private const val DELETE = "delete"
        private const val CHECK = "check"
    }

    fun manageModel(
        model: RemoteModel,
        call: MethodCall,
        result: MethodChannel.Result,
    ) {
        val task: String? = call.argument("task")

        if (task == null) {
            result.notImplemented()
            return
        }

        when (task) {
            DOWNLOAD -> {
                val isWifiReqRequired: Boolean = call.argument("wifi") ?: false
                val downloadConditions =
                    if (isWifiReqRequired) {
                        DownloadConditions.Builder().requireWifi().build()
                    } else {
                        DownloadConditions.Builder().build()
                    }
                downloadModel(model, downloadConditions, result)
            }

            DELETE -> {
                deleteModel(model, result)
            }

            CHECK -> {
                isModelDownloaded(
                    model,
                    object : CheckModelIsDownloadedCallback {
                        override fun onCheckResult(isDownloaded: Boolean?) {
                            result.success(isDownloaded)
                        }

                        override fun onError(e: Exception) {
                            result.error("error", e.toString(), null)
                        }
                    },
                )
            }

            else -> {
                result.notImplemented()
            }
        }
    }

    fun downloadModel(
        remoteModel: RemoteModel,
        downloadConditions: DownloadConditions,
        result: MethodChannel.Result,
    ) {
        isModelDownloaded(
            remoteModel,
            object : CheckModelIsDownloadedCallback {
                override fun onCheckResult(isDownloaded: Boolean?) {
                    if (isDownloaded == true) {
                        result.success("success")
                        return
                    }

                    remoteModelManager
                        .download(remoteModel, downloadConditions)
                        .addOnSuccessListener { result.success("success") }
                        .addOnFailureListener { exception -> result.error("error", exception.toString(), null) }
                }

                override fun onError(e: Exception) {
                    result.error("error", e.toString(), null)
                }
            },
        )
    }

    fun isModelDownloaded(
        model: RemoteModel,
        callback: CheckModelIsDownloadedCallback,
    ) {
        try {
            remoteModelManager
                .isModelDownloaded(model)
                .addOnFailureListener { exception -> callback.onError(exception) }
                .addOnSuccessListener { isModelDownloaded -> callback.onCheckResult(isModelDownloaded) }
        } catch (e: Exception) {
            callback.onError(e)
        }
    }

    fun deleteModel(
        remoteModel: RemoteModel,
        result: MethodChannel.Result,
    ) {
        isModelDownloaded(
            remoteModel,
            object : CheckModelIsDownloadedCallback {
                override fun onCheckResult(isDownloaded: Boolean?) {
                    if (isDownloaded != true) {
                        result.success("success")
                        return
                    }
                    remoteModelManager
                        .deleteDownloadedModel(remoteModel)
                        .addOnSuccessListener { result.success("success") }
                        .addOnFailureListener { exception -> result.error("error", exception.toString(), null) }
                }

                override fun onError(e: Exception) {
                    result.error("error", e.toString(), null)
                }
            },
        )
    }
}
